### DISCUSSION TITLE
useRouter returns undefined locale settings

### DISCUSSION THREAD

**Discussion by telriot** [vercel/next.js#18883] [Help] [answered]
Hi everyone!

I have been having trouble switching languages using next-translate as a localization library through the new 10.0.1 routing features as the ServerRouter return object from useRouter returns undefined parameters for the locales.
The next-config.js file has been set up for i18n according to the library examples, so I suppose the problem does not origin from there.

next-config.js:

    module.exports = {
	    i18n: {
		    locales: ["en", "ja-JP"],
		    defaultLocale: "en",
	    },
    };

This is the response I get when logging router from useRouter():

    ServerRouter {
    route: '/',
    pathname: '/',
    query: {},
    asPath: '/',
    basePath: '',
    events: undefined,
    isFallback: false,
    locale: undefined,
    locales: undefined,
    defaultLocale: undefined
}

and here is the component I am trying to use to switch languages:

    import Link from "next/link";
    import useTranslation from "next-translate/useTranslation";
    import i18nConfig from "../../i18n.json";
    import { useRouter } from "next/router";
    const { locales } = i18nConfig;

    function ChangeLanguage() {
	const { t, lang } = useTranslation();
	const router = useRouter();
	
    return (
		<>
			{locales.map((lng) => {
				if (lng === lang) return null;
				return (
					<Link href="/" locale={lng} key={`btn-change-${lng}`}>
						{t(`layout:language-name-${lng}`)}
					</Link>
				);
			})}
		</>
	    );
    }

I tried both using Link and useRouter with the same results.
Before thinking about a bug I was wondering if someone had similar experiences.

Thank you for your time and have a great day everybody!

**Comment by ijjk** [comment, marked as answer]
Hi, can you try the latest canary of Next.js `v10.0.2-canary.2`, this sounds related to some bug fixes that aren't in `v10.0.1`

**Reply by telriot** [reply]
Hi there, and thanks for the heads up!
Tried, but router parameters are still undefined and the switch does not work :/

**Reply by ijjk** [reply]
Can you provide a repo with a minimal reproduction where `useRouter().locale` is undefined with `i18n` support enabled?

**Reply by telriot** [reply]
Here it is:
https://github.com/telriot/next-router-bug

Please let me know if it works on your machine (it is my first time posting an issue and I may be doing things wrong!) and thank you!

**Reply by ijjk** [reply]
Hi, it looks like you have typo in the `next.config.js` filename it appears it's named `next-config.js` in the above repo which won't be recognized by Next.js, after updating the filename the locales appear to be loaded on `useRouter()` correctly.

**Reply by telriot** [reply]
Hi there!

That was indeed the case.
Sorry for having you waste your time looking through my typos :(
I have spent hours checking the code and I haven't taken a look at the filenames...

Everything is working now.

Thank you again, and have a great day!

Ben

**Comment by mlnima** [comment]
Hi , 

I have the same issue, the project is not single page application when the page load with no local "/" every link I click it will redirect and reload the page but when I switch to some other local except the default it works fine, only when there should be default locale, I get undefined https://github.com/mlnima/reactServerSideRenderingWithNext

**Reply by CopyJosh** [reply]
Having the same issue, seems to be introduced for me on 10.0.3-canary.0 where if the locale (undefined in this case) is not in the list of configured locales, the page does a window.location.href on navigation. Stepping through the router, the locales are defined, but defaultLocale is not.

**Comment by Writtscher** [comment]
Hi. Same issue for me. The locales are correct but the defaultLocale and the request locale is undefined.

```
module.exports = {
  i18n: {
    locales: ["de", "en", "es"], 
    defaultLocale: "de",
  },
};
```

When i inspect the router the locales are correct. But as mentioned the defaultLocale and the request locale is undefined.

```
{
  ...
  pathname: "/", 
  route: "/", 
  query: {…}, 
  asPath: "/",
  components: {…}, …},
  defaultLocale: undefined,
  locale: undefined
  locales: ["de", "en", "es"],
  ....
}
```

So I am stuck. Help is appreciated.

**Comment by CopyJosh** [comment]
My issue with pages not having a `locale` or `defaultLocale` set ended up being related to this one, since we were using a custom koa server and router. https://github.com/vercel/next.js/issues/20761

We had our routes set up as below, using the `next.render` method, which as described in the issue, does not run any of the `locale` utilities.
 
```js
router.get(['/:locale/foo', '/foo'], async (ctx) => {
  await next.render(ctx.request, ctx.res, '/foo', { ...ctx.query, ...ctx.params });
  return true;
});
```

I ended up copying some of the code from [/packages/next/next-server/server/next-server.ts](https://github.com/vercel/next.js/blob/0402fc459eb12a775372172a351f3a3f4fee83e3/packages/next/next-server/server/next-server.ts) into our own utility methods and adding the following to our router... (the `i18n` config was copied from `next.config.js` to our custom server config, which is what is being used as `config`.)

```js
router.use(async (ctx, next) => {
    // Try to set locale from url, session, header, or default
    const preferredLocale = utils.getLocale(ctx, config.i18n.locales)
      || utils.getLocaleFromSession(ctx)
      || utils.getLocaleFromHeader(ctx)
      || config.i18n.defaultLocale;

    ctx.state.locale = {
      __nextLocale: preferredLocale,
      __nextDefaultLocale: config.i18n.defaultLocale,
    };

    return next();
  });
```

New routes:

```js
router.get(['/:locale/foo', '/foo'], async (ctx) => {
  await next.render(ctx.request, ctx.res, '/foo', { ...ctx.query, ...ctx.params, ...ctx.state.locale });
  return true;
});
```

**Reply by Writtscher** [reply]
@CopyJosh 

Seems like same issue for me. I am using a custom nestjs server.
```
import { parse } from 'url';
import { Controller, Get, Res, Req } from '@nestjs/common';

import { Request, Response } from 'express';

import { NextServerProvider } from '../provider/next.provider';

@Controller({ path: '/' })
export class IndexController {
  constructor(private nextServerProvider: NextServerProvider) {}

  @Get()
  async showIndex(@Req() req: Request, @Res() res: Response) {
    const url = parse(req.url, true);
    await this.nextServerProvider
      .getNextServer()
      .render(req, res, url.pathname);
  }
}
```
 
This results in locales being undefined. But when I change it to this and call the `getRequestHandler` function, everything works fine.
```
import { Controller, Get, Res, Req } from '@nestjs/common';

import { Request, Response } from 'express';

import { NextServerProvider } from '../provider/next.provider';

@Controller({ path: '/' })
export class IndexController {
  constructor(private nextServerProvider: NextServerProvider) {}

  @Get()
  async showHome(@Req() req: Request, @Res() res: Response) {
    await this.nextServerProvider
      .getNextServer()
      .getRequestHandler()
      .call(req, res)
  }
}
```

So apparently this
```
...
     const server = next({ ... });
     await server.prepare();
     await server.render(....);
```

Does not work but
```
     const server = next({ ... });
     await server.prepare();
     await server.getRequestHandler().call(...);
```

Works. Learned it the hard way. Will have a look into this and see whether I can use your solution.

@ijjk Maybe you can help to understand the difference between the `render` and `getRequestHandler`.

**Comment by mlnima** [comment]
we can not use server.get for routes and the i18n at the same time, to overwrite the path use next.config settings, I really need to use custom routes on the express to cache the pages but since I am using I18n is not possible to cache the pages or add middleware for that route

**Comment by den4k-pr** [comment]
but if you import like this import { useRouter } from 'next/navigation';
then it is not possible to get locale, locales from the router in nextjs13.4
